Windows tests

sHaggYcaT opened the pull request:
**What this PR does / why we need it**:

The goal is implement 2 test cases for hyperv feature

**Special notes for your reviewer**:

ILpinto commented on tests/framework/util.go:
```diff
@@ -148,3 +137,72 @@ func RemoveDataVolume(dvName string, namespace string) {
+	if a < b {
+		return a
+	}
+	return b
```
use https://golang.org/pkg/math/#Min

sHaggYcaT replied:
OK, thank you

ILpinto commented on tests/framework/util.go:
```diff
@@ -148,3 +137,72 @@ func RemoveDataVolume(dvName string, namespace string) {
 	Expect(err).ToNot(HaveOccurred())
 }
+
+func CreateVMorVMIfromManifest(manifest, nameSpace string) string {
```
See tests/framework/types.go 
func (vm VirtualMachine) Create() (string, string, error)  does what you need

ILpinto commented on tests/framework/util.go:
```diff
@@ -148,3 +137,72 @@ func RemoveDataVolume(dvName string, namespace string) {
+
+// add by https://github.com/kubevirt/kubevirt-ansible/pull/531/, delete from here
+
+func GetLatestGitHubReleaseURL(user_name string, repo_name string) string {
```
why you need this ?

sHaggYcaT replied:
I already use it in several my test cases. Useful function to download latest release of any project in the github. For example, latest release of our common-templates. I can modify it and add another option - to download specific release. Already shown it to Martin.

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+	"encoding/xml"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	//	"reflect"
+)
```
Remove the unnecessary imports

sHaggYcaT replied:
OK

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+	return argument
+}
+
+func isVmiStarted(vmiName string) bool {
```
See tests/framework/types.go
use func (vm VirtualMachine) IsRunning() (bool, error)

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+	return result
+}
+
+func CreatingWinRmiPod(virtClient kubecli.KubevirtClient) *k8sv1.Pod {
```
move to utils

sHaggYcaT replied:
Sure. I planned to do this, but forget.

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+
+}
+
+func createPV() {
```
See tests/framework/util.go
Use func CreateResourceWithFilePathTestNamespace(filePath string)  you have PV yaml file.

sHaggYcaT replied:
I've deleted it

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+
+}
+
+func cleanPV(vmName string) {
```
See tests/framework/util.go
Use:
func DeleteResourceWithLabelTestNamespace(resourceType, resourceLabel string)

sHaggYcaT replied:
I've deleted it

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+
+// Checking if the cluster can run at least one VM
+// Todo: use this function to avoid overload CI
+func isEnoughResources(virtClient kubecli.KubevirtClient, cpuNeeded int, memNeeded int64) (bool, int, int) {
```
move to utils

ILpinto commented on tests/win_test.go:
```diff
@@ -0,0 +1,370 @@
+		It("testcase", func() {
+
+			var vmiHyperv tframework.VirtualMachine
+			vmiHyperv.Manifest = address_common + "/hyperv.yml"
```
hyperv.yaml is missing

sHaggYcaT replied:
ooops, sorry

ILpinto commented:
Add also polarion test ids

fromanirh commented on tests/win_test.go:
```diff
@@ -338,20 +357,83 @@ fmt.Scanln(&input1)
+		Expect(err).ToNot(HaveOccurred())
+			
+		By("Checking HyperV emulation features for KVM")	
+		features := domStat.Features
```
this may be nil

sHaggYcaT replied:
I've changed it

fromanirh commented on tests/win_test.go:
```diff
@@ -338,20 +357,83 @@ fmt.Scanln(&input1)
+			
+		By("Checking HyperV emulation features for KVM")	
+		features := domStat.Features
+               	hyperv := features.Hyperv
```
also this may be nil. In this case the test will crash, but I guess that could count as a failure

sHaggYcaT replied:
I've changed it

fromanirh commented on tests/win_test.go:
```diff
@@ -338,20 +357,83 @@ fmt.Scanln(&input1)
+			case "hpet":
+				Expect(Timer[i].Present).To(Equal("no"))
+
+			case "hypervclock":
```
I think you should just check this and just ignore everything else

sHaggYcaT replied:
I've changed it

ILpinto commented on tests/framework/util.go:
```diff
@@ -18,6 +19,8 @@ import (
 	//delete in PR
```
? remove this

ILpinto commented on tests/framework/util.go:
```diff
@@ -18,6 +19,8 @@ import (
 	//delete in PR
 	"net/http"
 	"os/exec"
+//	"k8s.io/apimachinery/pkg/util/rand"
```
If not in use remove it

ILpinto commented on tests/framework/types.go:
```diff
@@ -39,12 +39,12 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Create() (string, string, error) {
 	args := []string{"create", "-f", vm.Manifest}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtOcPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace,"oc", args...)
```
Use the path to oc and virtctl are path of run parameters.

sHaggYcaT replied:
I've copied solution from my another PR

ILpinto commented on tests/framework/types.go:
```diff
@@ -39,12 +39,12 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Start() (string, string, error) {
 	args := []string{"start", vm.Name}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtVirtctlPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace,"virtctl", args...)
```
same

sHaggYcaT replied:
I've copied solution from my another PR

sHaggYcaT commented:
I've deleted from this PR Windows part, and implemented @fromanirh suggestions for first test case, also I've wrote new one test case (to test hyperv feature inside VM).
Please review it.
Note: it uses my own docker image. I'm going to ask to add it in our repo.

fromanirh commented on tests/hyperv_test.go:
```diff
@@ -0,0 +1,259 @@
+const (
+	windowsDisk         = "windows-disk"
+	username            = "Administrator"
+	password            = "Heslo123"
```
I wonder if there is a better way than hardcode passwords in the source code

sHaggYcaT replied:
Agree, but same way uses in the unit tests. Probably it's not important image. For new test framework we can write some kind of *infrastructure to store secrets*

fromanirh commented on tests/hyperv_test.go:
```diff
@@ -0,0 +1,259 @@
+	vmi.Type = "vmi"
+	const maxWaitIterations = 60 // two minutes
+	for i := 0; i < maxWaitIterations; i++ {
+		phaseStatus, err := vmi.IsRunning()
```
minor: misleading name, should be something like "isRunning" or "vmRunning" or something like that
